File: doc/json/query-docs.py
#!/usr/bin/env python3
"""
JSDoc JSON Documentation Query Tool
Python version for querying Snap.svg documentation

Usage:
    python query-docs.py --class Element
    python query-docs.py --method animate --class Element
    python query-docs.py --search animate
"""

import json
import argparse
import sys
from pathlib import Path


class DocQuery:
    def __init__(self, doc_file='documentation.json'):
        """Load the documentation JSON file"""
        doc_path = Path(__file__).parent / doc_file
        with open(doc_path, 'r', encoding='utf-8') as f:
            self.docs = json.load(f)

    def find_class(self, class_name):
        """Find a class by name"""
        for item in self.docs:
            if item.get('kind') == 'class' and \
               (item.get('name') == class_name or item.get('longname') == class_name):
                return item
        return None

    def find_class_methods(self, class_name):
        """Find all methods of a class"""
        methods = []
        for item in self.docs:
            if item.get('kind') == 'function' and item.get('memberof') == class_name:
                methods.append(item)
        return methods

    def search_by_name(self, pattern):
        """Search for items by name pattern (case-insensitive)"""
        pattern_lower = pattern.lower()
        results = []
        for item in self.docs:
            name = item.get('name', '')
            longname = item.get('longname', '')
            if pattern_lower in name.lower() or pattern_lower in longname.lower():
                results.append(item)
        return results

    def get_all_classes(self):
        """Get all documented classes"""
        return [item for item in self.docs if item.get('kind') == 'class']

    def get_all_functions(self):
        """Get all functions/methods"""
        return [item for item in self.docs if item.get('kind') == 'function']

    def get_comment(self, doc_item):
        """Get the description/comment for a documentation item"""
        return doc_item.get('description') or doc_item.get('comment') or 'No documentation available'

    def print_item(self, item, verbose=False):
        """Pretty print a documentation item"""
        name = item.get('name', 'Unknown')
        kind = item.get('kind', 'unknown')
        longname = item.get('longname', name)

        print(f"\n{'='*60}")
        print(f"Name: {name}")
        print(f"Full Name: {longname}")
        print(f"Kind: {kind}")

        if item.get('memberof'):
            print(f"Member of: {item['memberof']}")

        if verbose:
            description = self.get_comment(item)
            print(f"\nDescription:\n{description[:200]}...")

            if item.get('params'):
                print(f"\nParameters:")
                for param in item['params']:
                    param_name = param.get('name', 'unknown')
                    param_type = param.get('type', {}).get('names', ['unknown'])
                    param_desc = param.get('description', '')
                    print(f"  - {param_name} ({', '.join(param_type)}): {param_desc}")

            if item.get('returns'):
                print(f"\nReturns:")
                for ret in item['returns']:
                    ret_type = ret.get('type', {}).get('names', ['unknown'])
                    ret_desc = ret.get('description', '')
                    print(f"  - ({', '.join(ret_type)}): {ret_desc}")

            if item.get('meta'):
                meta = item['meta']
                print(f"\nSource: {meta.get('filename')}:{meta.get('lineno')}")


def main():
    parser = argparse.ArgumentParser(description='Query JSDoc JSON documentation')
    parser.add_argument('--class', dest='class_name', help='Find a specific class')
    parser.add_argument('--method', dest='method_name', help='Find a specific method (use with --class)')
    parser.add_argument('--search', help='Search for items by name pattern')
    parser.add_argument('--list-classes', action='store_true', help='List all classes')
    parser.add_argument('--list-methods', action='store_true', help='List all methods of a class (use with --class)')
    parser.add_argument('-v', '--verbose', action='store_true', help='Show detailed information')

    args = parser.parse_args()

    # Load documentation
    try:
        query = DocQuery()
    except FileNotFoundError:
        print("Error: documentation.json not found. Run 'grunt docs:json' first.", file=sys.stderr)
        sys.exit(1)

    # Process commands
    if args.class_name:
        cls = query.find_class(args.class_name)
        if cls:
            query.print_item(cls, args.verbose)

            if args.list_methods:
                methods = query.find_class_methods(args.class_name)
                print(f"\n\nMethods of {args.class_name} ({len(methods)} total):")
                for method in methods[:20]:  # Show first 20
                    print(f"  - {method.get('name')}()")
                if len(methods) > 20:
                    print(f"  ... and {len(methods) - 20} more")

            if args.method_name:
                methods = query.find_class_methods(args.class_name)
                method = next((m for m in methods if m.get('name') == args.method_name), None)
                if method:
                    print(f"\n\nMethod Details:")
                    query.print_item(method, verbose=True)
                else:
                    print(f"\nMethod '{args.method_name}' not found in class '{args.class_name}'")
        else:
            print(f"Class '{args.class_name}' not found")

    elif args.search:
        results = query.search_by_name(args.search)
        print(f"Found {len(results)} items matching '{args.search}':")
        for item in results[:20]:  # Show first 20
            query.print_item(item, args.verbose)
        if len(results) > 20:
            print(f"\n... and {len(results) - 20} more results")

    elif args.list_classes:
        classes = query.get_all_classes()
        print(f"All Classes ({len(classes)} total):")
        for cls in classes:
            print(f"  - {cls.get('name', 'Unknown')} ({cls.get('longname', '')})")

    else:
        parser.print_help()
        print("\n\nExamples:")
        print("  python query-docs.py --class Element -v")
        print("  python query-docs.py --class Element --method animate")
        print("  python query-docs.py --class Element --list-methods")
        print("  python query-docs.py --search animate -v")
        print("  python query-docs.py --list-classes")


if __name__ == '__main__':
    main()

File: utils/introspection/query-docs.js
/**
 * Snap.svg Documentation Reader - Node.js CLI
 *
 * Command-line tool for querying JSDoc documentation
 *
 * Usage:
 *   node query-docs.js --class Element
 *   node query-docs.js --class Element --method attr
 *   node query-docs.js --search "animate"
 */

const fs = require('fs');
const path = require('path');
const SnapDocsReader = require('./snap-docs-reader.js');

// Load documentation from file
function loadDocs() {
    const docPath = path.resolve(__dirname, '../../doc/json/documentation.json');

    if (!fs.existsSync(docPath)) {
        console.error('Error: documentation.json not found at:', docPath);
        console.error('Run "grunt docs:json" to generate it.');
        process.exit(1);
    }

    const data = JSON.parse(fs.readFileSync(docPath, 'utf8'));
    return SnapDocsReader.load(data);
}

// Format parameter for display
function formatParam(param) {
    let str = param.name;
    if (param.optional) str = `[${str}]`;
    if (param.type && param.type.names) {
        str += `: ${param.type.names.join('|')}`;
    }
    if (param.description) {
        str += ` - ${param.description}`;
    }
    return str;
}

// Display documentation for a class
function displayClass(className, verbose = false) {
    const classDoc = SnapDocsReader.findClass(className);

    if (!classDoc) {
        console.log(`Class "${className}" not found in documentation.`);
        return;
    }

    console.log('\n' + '='.repeat(60));
    console.log(`Class: ${className}`);
    console.log('='.repeat(60));

    if (classDoc.description) {
        console.log('\nDescription:');
        console.log(classDoc.description);
    }

    if (verbose && classDoc.comment) {
        console.log('\nFull Comment:');
        console.log(classDoc.comment);
    }

    // Show methods
    const methods = SnapDocsReader.findClassMethods(className);
    if (methods.length > 0) {
        console.log(`\nMethods (${methods.length}):`);
        methods.forEach(method => {
            const sig = SnapDocsReader.getSignature(method);
            console.log(`  - ${sig}`);
            if (verbose && method.description) {
                console.log(`    ${method.description}`);
            }
        });
    }
}

// Display documentation for a method
function displayMethod(className, methodName, verbose = false) {
    const methodDoc = SnapDocsReader.findMethod(className, methodName);

    if (!methodDoc) {
        console.log(`Method "${methodName}" not found in class "${className}".`);
        return;
    }

    console.log('\n' + '='.repeat(60));
    console.log(`${className}.${methodName}`);
    console.log('='.repeat(60));

    console.log('\nSignature:');
    console.log('  ' + SnapDocsReader.getSignature(methodDoc));

    if (methodDoc.description) {
        console.log('\nDescription:');
        console.log('  ' + methodDoc.description);
    }

    const params = SnapDocsReader.getParams(methodDoc);
    if (params.length > 0) {
        console.log('\nParameters:');
        params.forEach(param => {
            console.log('  - ' + formatParam(param));
        });
    }

    const returns = SnapDocsReader.getReturns(methodDoc);
    if (returns) {
        console.log('\nReturns:');
        if (returns.type && returns.type.names) {
            console.log('  Type: ' + returns.type.names.join('|'));
        }
        if (returns.description) {
            console.log('  ' + returns.description);
        }
    }

    if (verbose && methodDoc.comment) {
        console.log('\nFull JSDoc Comment:');
        console.log(methodDoc.comment);
    }
}

// Search documentation
function displaySearch(query, verbose = false) {
    const results = SnapDocsReader.search(query);

    console.log(`\nSearch results for "${query}": ${results.length} matches\n`);
    console.log('='.repeat(60));

    results.slice(0, 20).forEach(item => {
        console.log(`\n${item.longname || item.name}`);
        console.log(`  Kind: ${item.kind}`);
        if (item.description) {
            const desc = item.description.substring(0, 100);
            console.log(`  ${desc}${item.description.length > 100 ? '...' : ''}`);
        }
        if (verbose) {
            const sig = SnapDocsReader.getSignature(item);
            if (sig) {
                console.log(`  Signature: ${sig}`);
            }
        }
    });

    if (results.length > 20) {
        console.log(`\n... and ${results.length - 20} more results`);
    }
}

// List all classes
function listClasses() {
    const classes = SnapDocsReader.getAllClasses();

    console.log(`\nDocumented Classes (${classes.length}):\n`);
    console.log('='.repeat(60));

    classes.forEach(cls => {
        console.log(`\n${cls.name}`);
        if (cls.description) {
            const desc = cls.description.substring(0, 80);
            console.log(`  ${desc}${cls.description.length > 80 ? '...' : ''}`);
        }
        const methods = SnapDocsReader.findClassMethods(cls.name);
        console.log(`  Methods: ${methods.length}`);
    });
}

// Parse command-line arguments
function parseArgs() {
    const args = process.argv.slice(2);
    const options = {
        className: null,
        methodName: null,
        search: null,
        listClasses: false,
        listMethods: false,
        verbose: false
    };

    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--class':
            case '-c':
                options.className = args[++i];
                break;
            case '--method':
            case '-m':
                options.methodName = args[++i];
                break;
            case '--search':
            case '-s':
                options.search = args[++i];
                break;
            case '--list-classes':
                options.listClasses = true;
                break;
            case '--list-methods':
                options.listMethods = true;
                break;
            case '--verbose':
            case '-v':
                options.verbose = true;
                break;
            case '--help':
            case '-h':
                showHelp();
                process.exit(0);
        }
    }

    return options;
}

// Show help message
function showHelp() {
    console.log(`
Snap.svg Documentation Query Tool

Usage:
  node query-docs.js [options]

Options:
  --class, -c <name>       Show documentation for a class
  --method, -m <name>      Show documentation for a method (requires --class)
  --search, -s <query>     Search documentation
  --list-classes           List all documented classes
  --list-methods           List methods of a class (requires --class)
  --verbose, -v            Show verbose output
  --help, -h               Show this help message

Examples:
  node query-docs.js --class Element
  node query-docs.js --class Element --method attr
  node query-docs.js --class Element --list-methods
  node query-docs.js --search "animate"
  node query-docs.js --list-classes
`);
}

// Main execution
function main() {
    const options = parseArgs();

    // Show help if no options
    if (!options.className && !options.search && !options.listClasses) {
        showHelp();
        return;
    }

    // Load documentation
    loadDocs();

    // Execute command
    if (options.listClasses) {
        listClasses();
    } else if (options.search) {
        displaySearch(options.search, options.verbose);
    } else if (options.className) {
        if (options.methodName) {
            displayMethod(options.className, options.methodName, options.verbose);
        } else if (options.listMethods) {
            const methods = SnapDocsReader.findClassMethods(options.className);
            console.log(`\nMethods of ${options.className} (${methods.length}):\n`);
            methods.forEach(method => {
                console.log('  ' + SnapDocsReader.getSignature(method));
            });
        } else {
            displayClass(options.className, options.verbose);
        }
    }

    console.log('');
}

// Run if called directly
if (require.main === module) {
    main();
}

// Export for programmatic use
module.exports = {
    loadDocs,
    displayClass,
    displayMethod,
    displaySearch,
    listClasses,
    SnapDocsReader
};

